feat: Add div mod hint

coxmars opened the pull request:
Hi @zmalatrax,

I've made significant progress on the `DivMod` hint implementation, including updates to the `hintName.ts`, `hintSchema.ts`, `hintHandler.ts`. The core logic is nearly complete, but a few final adjustments are still needed in `divMod.ts`. Also, the `verified` tag is working according with the contribution guidelines.

#### Status
- Add div mod hint working as expected.

coxmars commented:
#### Update 📌

Hi @zmalatrax

- Implemented and tested the `DivMod` hint.
- All related tests for `DivMod` have passed successfully.

- Here are the test results:
  
<img width="413" alt="image" src="https://github.com/user-attachments/assets/d0bc6d33-b512-4bcc-98f0-b3b594442c83">

- Please review the changes and let me know if any additional modifications are needed, also can you explain me a little bit the integration test with Cairo I think it is the only I need to complete it 🫡

zmalatrax commented on src/hints/math/divMod.ts:
```diff
@@ -0,0 +1,65 @@
+    const quotientValue = new Felt(lhsValue / rhsValue);
+    const remainderValue = new Felt(lhsValue % rhsValue);
+    
+    const quotientAddr = vm.cellRefToRelocatable(quotientAddress);
+    const remainderAddr = vm.cellRefToRelocatable(remainderAddress);
+
+    vm.memory.assertEq(quotientAddr, quotientValue);
+    vm.memory.assertEq(remainderAddr, remainderValue);
```
No need to declare new variables for quotient and remainder addresses, you can directly call `vm.memory.assertEq(vm.cellRefToRelocatable(quotient), quotientValue)`

zmalatrax commented on src/hints/math/divMod.test.ts:
```diff
@@ -0,0 +1,92 @@
+    const quotientExpected = new Felt(3n);
+    const remainderExpected = new Felt(2n);
+
+    const lhsAddr = new Relocatable(0, 0);
+    const rhsAddr = new Relocatable(0, 1);
+    const quotientAddr = new Relocatable(0, 2);
+    const remainderAddr = new Relocatable(0, 3);
+
+    vm.memory.assertEq(lhsAddr, lhsValue);
+    vm.memory.assertEq(rhsAddr, rhsValue);
+    vm.ap = new Relocatable(0, 0);  
```
Why storing the values in the segment 0 while `ap` points to the segment 1?

Leave `ap` as is (it points to `Relocatable(1,0) by default)

Use `vm.ap.add()` to compute the addresses

```suggestion
vm.memory.assertEq(vm.ap, lhsValue);
vm.memory.assertEq(vm.ap.add(1), rhsValue);
```

zmalatrax commented on src/hints/math/divMod.test.ts:
```diff
@@ -0,0 +1,92 @@
+
+    divMod(vm, hint.leftHandSide, hint.rightHandSide, hint.quotientAddress, hint.remainderAddress);
+
+    expect(vm.memory.get(quotientAddr)).toEqual(quotientExpected);
+    expect(vm.memory.get(remainderAddr)).toEqual(remainderExpected);
```
use `vm.cellRefToRelocatable(hint.quotient)` rather than `quotientAddr`, same for the remainder

zmalatrax commented on src/hints/math/divMod.test.ts:
```diff
@@ -0,0 +1,92 @@
+    vm.memory.addSegment();
+    vm.memory.addSegment();
+
+    const lhsValue = new Felt(17n);
+    const rhsValue = new Felt(5n);
+    const quotientExpected = new Felt(3n);
+    const remainderExpected = new Felt(2n);
```
Make a few different cases by parameterizing the test, it can be done with `test.each`, the [TestLessThan test](https://github.com/kkrt-labs/cairo-vm-ts/blob/e133561a28ea1f269f6efd1af8bb657007c20b74/src/hints/math/testLessThan.test.ts#L53-L58) can be used as a reference
And here is the [doc](https://bun.sh/docs/test/writing#test-each)

zmalatrax reviewed (commented):
great :)

coxmars commented:
> @coxmars Looks like your typescript files aren't well formatted, we're using Trunk, I'd recommend you installing their [extension](https://docs.trunk.io/code-quality/ide-integration/vs-code) if you're on VSCode or Neovim
> 
> Otherwise, you can add their [CLI ](https://docs.trunk.io/code-quality/usage) and run `trunk fmt`.

Thanks for the feedback, I will apply the requested changes asap 🫡

zmalatrax commented:
About the integration test, you need to add a Cairo program `divmod.cairo` in the `cairo_programs/cairo/hints` folder that makes use of the DivMod hint.

It should be a simple program focused on this hint, such as
```cairo
fn main() {
    let _ = 17_u32 / 12_u32;
    let _ = 17_u32 % 12_u32;
}
```

coxmars commented:
Hi @zmalatrax I'm trying to execute make compile and I got this error:

<img width="764" alt="image" src="https://github.com/user-attachments/assets/9301bf56-33cc-47c2-8ad2-3e984515cd0d">

- I already have Bun, Poetry, I ran poetry install for dependencies and I downloaded cairo-lang too and also Python 3.10 so do you have an idea what is the problem?

coxmars commented:
Hi @zmalatrax I was able to solve some problems running the integration test haha, it took me a while but I got it done, unit test and integration test ready plus the feedback given, as you can see below it is working so I will open this draft PR to "Ready for review" 🫡

<img width="856" alt="image" src="https://github.com/user-attachments/assets/44b224a8-abdd-44f7-a6c5-e4c353947480">

zmalatrax commented on src/hints/math/divMod.ts:
```diff
@@ -0,0 +1,66 @@
+  const lhsValue = vm.getResOperandValue(lhs).toBigInt();
+  const rhsValue = vm.getResOperandValue(rhs).toBigInt();
+
+  if (rhsValue === 0n) {
+    throw new Error('Division by zero');
+  }
```
Use `!rhsValue` instead of `rhsValue === 0n` in the if condition

Use the `CannotDivideByZero()` error defined in `errors/primitives.ts`

Avoid using brackets when it is a simple if condition which throws an error

```suggestion
if (!rhsValue) throw new CannotDivideByZero()`
```

enitrat commented on src/hints/math/divMod.test.ts:
```diff
@@ -0,0 +1,91 @@
+    });
+  });
+
+  test.each([
```
test cases are not covering edge cases

coxmars commented:
Hi @zmalatrax I already changed the condition as you said and @enitrat I added a new edge case test, if you need something else to be improved just tell me, thanks 🫡

zmalatrax commented on src/hints/math/divMod.test.ts:
```diff
@@ -88,4 +88,21 @@ describe('DivMod', () => {
+    vm.memory.assertEq(vm.ap, lhsValue);
+    vm.memory.assertEq(vm.ap.add(1), rhsValue);
+
+    expect(() => {
+      divMod(vm, hint.lhs, hint.rhs, hint.quotient, hint.remainder);
+    });
```
You're expecting the  `divMod` function to throw the error `CannotDivideByZero()`, it is not reflected in the test

It should be of the form `expect(() => foo()).toThrow(new MyError())`

coxmars commented:
Hi @zmalatrax I applied the changes and add a few more tests ✅

<img width="424" alt="image" src="https://github.com/user-attachments/assets/01a5a927-0cd9-42ce-9422-bfb5106a92fc">

<img width="474" alt="image" src="https://github.com/user-attachments/assets/03469a91-284f-403e-b3ac-0af35c575bfc">

zmalatrax commented on src/hints/math/divMod.test.ts:
```diff
@@ -0,0 +1,111 @@
+
+    expect(() => {
+      divMod(vm, hint.lhs, hint.rhs, hint.quotient, hint.remainder);
+    }).toThrow(CannotDivideByZero);
```
We test that the error message are similar in the codebase, using an instance of the error rather than testing that the instance of the class are smilar, see [Jest documentation](https://jestjs.io/docs/expect#tothrowerror) for reference

```suggestion
    }).toThrow(new CannotDivideByZero());
```

coxmars commented:
Hi @zmalatrax, I already applied the improvement, thanks 🫡

zmalatrax reviewed (approved):
good
